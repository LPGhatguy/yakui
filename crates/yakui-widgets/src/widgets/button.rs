use std::borrow::Cow;

use yakui_core::event::{EventInterest, EventResponse, WidgetEvent};
use yakui_core::geometry::Color3;
use yakui_core::input::MouseButton;
use yakui_core::widget::Widget;
use yakui_core::{Alignment, Response};

use crate::colors;
use crate::style::{TextAlignment, TextStyle};
use crate::util::widget;
use crate::widgets::Pad;

use super::RenderText;

/**
A button containing some text.

Responds with [ButtonResponse].

Shorthand:
```rust
# let _handle = yakui_widgets::DocTest::start();
if yakui::button("Hello").clicked {
    println!("The button was clicked");
}
```
*/
#[derive(Debug)]
#[non_exhaustive]
pub struct Button {
    pub text: Cow<'static, str>,
    pub text_style: TextStyle,
    pub padding: Pad,
    pub fill: Color3,
    pub hover_fill: Option<Color3>,
    pub down_fill: Option<Color3>,
}

impl Button {
    pub fn unstyled(text: Cow<'static, str>) -> Self {
        let mut text_style = TextStyle::label();
        text_style.align = TextAlignment::Center;

        Self {
            text,
            text_style,
            padding: Pad::ZERO,
            fill: Color3::GRAY,
            hover_fill: None,
            down_fill: None,
        }
    }

    pub fn styled(text: Cow<'static, str>) -> Self {
        let mut text_style = TextStyle::label();
        text_style.align = TextAlignment::Center;

        Self {
            text,
            text_style,
            padding: Pad::balanced(20.0, 10.0),
            fill: colors::BACKGROUND_3,
            hover_fill: Some(colors::BACKGROUND_3.adjust(1.2)),
            down_fill: Some(colors::BACKGROUND_3.adjust(0.8)),
        }
    }

    pub fn show(self) -> Response<ButtonWidget> {
        widget::<ButtonWidget>(self)
    }
}

#[derive(Debug)]
pub struct ButtonWidget {
    props: Button,
    hovering: bool,
    mouse_down: bool,
    clicked: bool,
}

#[derive(Debug)]
pub struct ButtonResponse {
    pub hovering: bool,
    pub clicked: bool,
}

impl Widget for ButtonWidget {
    type Props = Button;
    type Response = ButtonResponse;

    fn new() -> Self {
        Self {
            props: Button::unstyled(Cow::Borrowed("")),
            hovering: false,
            mouse_down: false,
            clicked: false,
        }
    }

    fn update(&mut self, props: Self::Props) -> Self::Response {
        self.props = props;

        let mut color = self.props.fill;

        if let (Some(fill), true) = (self.props.down_fill, self.mouse_down) {
            color = fill
        } else if let (Some(hover), true) = (self.props.hover_fill, self.hovering) {
            color = hover
        }

        let alignment = match self.props.text_style.align {
            TextAlignment::Start => Alignment::CENTER_LEFT,
            TextAlignment::Center => Alignment::CENTER,
            TextAlignment::End => Alignment::CENTER_RIGHT,
        };

        crate::colored_box_container(color, || {
            crate::pad(self.props.padding, || {
                crate::align(alignment, || {
                    let mut text = RenderText::label(self.props.text.clone());
                    text.style = self.props.text_style.clone();
                    text.show();
                });
            });
        });

        let clicked = self.clicked;
        self.clicked = false;

        Self::Response {
            hovering: self.hovering,
            clicked,
        }
    }

    fn event_interest(&self) -> EventInterest {
        EventInterest::MOUSE
    }

    fn event(&mut self, event: &WidgetEvent) -> EventResponse {
        match event {
            WidgetEvent::MouseEnter => {
                self.hovering = true;
                EventResponse::Sink
            }
            WidgetEvent::MouseLeave => {
                self.hovering = false;
                EventResponse::Sink
            }
            WidgetEvent::MouseButtonChanged(MouseButton::One, down) => {
                if *down {
                    self.mouse_down = true;
                    EventResponse::Sink
                } else if self.mouse_down {
                    self.mouse_down = false;
                    self.clicked = true;
                    EventResponse::Sink
                } else {
                    EventResponse::Bubble
                }
            }
            WidgetEvent::MouseButtonChangedOutside(MouseButton::One, down) => {
                if !*down {
                    self.mouse_down = false;
                }

                EventResponse::Bubble
            }
            _ => EventResponse::Bubble,
        }
    }
}
