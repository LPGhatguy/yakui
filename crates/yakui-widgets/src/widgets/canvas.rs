use yakui_core::widget::{PaintContext, Widget};
use yakui_core::Response;

use crate::ignore_debug::IgnoreDebug;
use crate::util::{widget, widget_children};

type DrawCallback = Box<dyn Fn(&mut PaintContext<'_>) + 'static>;

/**
Allows the user to draw arbitrary graphics in a region.

Responds with [CanvasResponse].
*/
#[derive(Debug)]
pub struct Canvas {
    draw: IgnoreDebug<Option<DrawCallback>>,
}

impl Canvas {
    pub fn new(draw: impl Fn(&mut PaintContext<'_>) + 'static) -> Self {
        Self {
            draw: IgnoreDebug(Some(Box::new(draw))),
        }
    }

    pub fn show(self) -> Response<CanvasResponse> {
        widget::<CanvasWidget>(self)
    }

    pub fn show_children<F: FnOnce()>(self, children: F) -> Response<CanvasResponse> {
        widget_children::<CanvasWidget, F>(children, self)
    }
}

#[derive(Debug)]
pub struct CanvasWidget {
    props: Canvas,
}

pub type CanvasResponse = ();

impl Widget for CanvasWidget {
    type Props = Canvas;
    type Response = CanvasResponse;

    fn new() -> Self {
        Self {
            props: Canvas {
                draw: IgnoreDebug(None),
            },
        }
    }

    fn update(&mut self, props: Self::Props) -> Self::Response {
        self.props = props;
    }

    fn paint(&self, mut ctx: PaintContext<'_>) {
        if let IgnoreDebug(Some(draw)) = &self.props.draw {
            (draw)(&mut ctx);
        }

        self.default_paint(ctx);
    }
}
