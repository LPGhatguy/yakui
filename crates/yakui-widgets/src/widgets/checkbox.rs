use yakui_core::event::{EventInterest, EventResponse, WidgetEvent};
use yakui_core::geometry::{Constraints, Vec2};
use yakui_core::input::MouseButton;
use yakui_core::paint::PaintRect;
use yakui_core::widget::{EventContext, LayoutContext, PaintContext, Widget};
use yakui_core::Response;

use crate::colors;

const OUTER_SIZE: f32 = 24.0;
const INNER_SIZE: f32 = 16.0;

/**
A checkbox with a provided value.

Responds with [CheckboxResponse].

Shorthand:
```rust
# let _handle = yakui_widgets::DocTest::start();
let mut value = false;

value = yakui::checkbox(value).checked;
```
*/
#[derive(Debug)]
#[non_exhaustive]
pub struct Checkbox {
    pub checked: bool,
}

impl Checkbox {
    pub fn new(checked: bool) -> Self {
        Self { checked }
    }

    pub fn show(self) -> Response<CheckboxWidget> {
        crate::util::widget::<CheckboxWidget>(self)
    }
}

#[derive(Debug)]
pub struct CheckboxWidget {
    props: Checkbox,
    hovering: bool,
    mouse_down: bool,
    just_toggled: bool,
}

#[derive(Debug)]
#[non_exhaustive]
pub struct CheckboxResponse {
    pub checked: bool,
}

impl Widget for CheckboxWidget {
    type Props = Checkbox;
    type Response = CheckboxResponse;

    fn new() -> Self {
        Self {
            props: Checkbox::new(false),
            hovering: false,
            mouse_down: false,
            just_toggled: false,
        }
    }

    fn update(&mut self, props: Self::Props) -> Self::Response {
        self.props = props;

        let mut checked = self.props.checked;
        if self.just_toggled {
            checked = !checked;
            self.just_toggled = false;
        }

        CheckboxResponse { checked }
    }

    fn paint(&self, ctx: PaintContext<'_>) {
        let layout_node = ctx.layout.get(ctx.dom.current()).unwrap();

        let padding = Vec2::splat(OUTER_SIZE - INNER_SIZE);
        let mut check_rect = layout_node.rect;
        check_rect.set_pos(check_rect.pos() + padding / 2.0);
        check_rect.set_size(check_rect.size() - padding);

        let mut bg = PaintRect::new(layout_node.rect);
        bg.color = colors::BACKGROUND_3;
        ctx.paint.add_rect(bg);

        if self.props.checked {
            crate::icons::cross(ctx.paint, check_rect, colors::TEXT);
        }
    }

    fn layout(&self, _ctx: LayoutContext<'_>, constraints: Constraints) -> Vec2 {
        constraints.constrain_min(Vec2::splat(OUTER_SIZE))
    }

    fn event_interest(&self) -> EventInterest {
        EventInterest::MOUSE_INSIDE | EventInterest::MOUSE_OUTSIDE
    }

    fn event(&mut self, _ctx: EventContext<'_>, event: &WidgetEvent) -> EventResponse {
        match event {
            WidgetEvent::MouseEnter => {
                self.hovering = true;
                EventResponse::Sink
            }
            WidgetEvent::MouseLeave => {
                self.hovering = false;
                EventResponse::Sink
            }
            WidgetEvent::MouseButtonChanged {
                button: MouseButton::One,
                down,
                inside,
                ..
            } => {
                if *inside {
                    if *down {
                        self.mouse_down = true;
                        EventResponse::Sink
                    } else if self.mouse_down {
                        self.mouse_down = false;
                        self.just_toggled = true;
                        EventResponse::Sink
                    } else {
                        EventResponse::Bubble
                    }
                } else {
                    self.mouse_down = false;
                    EventResponse::Bubble
                }
            }
            _ => EventResponse::Bubble,
        }
    }
}
