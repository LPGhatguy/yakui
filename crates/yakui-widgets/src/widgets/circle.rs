use yakui_core::geometry::{Color, Constraints, Vec2};
use yakui_core::widget::{LayoutContext, PaintContext, Widget};
use yakui_core::Response;

use crate::shapes;
use crate::util::{widget, widget_children};

/**
A colored circle that can contain children.

Responds with [CircleResponse].
*/
#[derive(Debug, Clone)]
#[non_exhaustive]
pub struct Circle {
    pub color: Color,
    pub min_radius: f32,
}

impl Circle {
    pub fn new() -> Self {
        Self {
            color: Color::WHITE,
            min_radius: 0.0,
        }
    }

    pub fn show(self) -> Response<CircleResponse> {
        widget::<CircleWidget>(self)
    }

    pub fn show_children<F: FnOnce()>(self, children: F) -> Response<CircleResponse> {
        widget_children::<CircleWidget, F>(children, self)
    }
}

#[derive(Debug)]
pub struct CircleWidget {
    props: Circle,
}

pub type CircleResponse = ();

impl Widget for CircleWidget {
    type Props = Circle;
    type Response = CircleResponse;

    fn new() -> Self {
        Self {
            props: Circle::new(),
        }
    }

    fn update(&mut self, props: Self::Props) -> Self::Response {
        self.props = props;
    }

    fn layout(&self, mut ctx: LayoutContext<'_>, input: Constraints) -> Vec2 {
        let node = ctx.dom.get_current();
        let mut size = Vec2::splat(self.props.min_radius);

        for &child in &node.children {
            let child_size = ctx.calculate_layout(child, input);
            size = size.max(child_size);
        }

        input.constrain_min(size)
    }

    fn paint(&self, mut ctx: PaintContext<'_>) {
        let node = ctx.dom.get_current();
        let layout_node = ctx.layout.get(ctx.dom.current()).unwrap();

        let center = layout_node.rect.pos() + layout_node.rect.size() / 2.0;
        let radius = layout_node.rect.size().x.min(layout_node.rect.size().y) / 2.0;
        let mut circle = shapes::Circle::new(center, radius);
        circle.color = self.props.color;
        circle.add(ctx.paint);

        for &child in &node.children {
            ctx.paint(child);
        }
    }
}
