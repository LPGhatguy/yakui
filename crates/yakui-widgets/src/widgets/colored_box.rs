use yakui_core::geometry::{Color, Constraints, Vec2};
use yakui_core::paint::PaintRect;
use yakui_core::widget::{LayoutContext, PaintContext, Widget};
use yakui_core::Response;

use crate::util::{widget, widget_children};

/**
A colored box that can contain children.

Responds with [ColoredBoxResponse].
*/
#[derive(Debug, Clone)]
#[non_exhaustive]
pub struct ColoredBox {
    pub color: Color,
    pub min_size: Vec2,
}

impl ColoredBox {
    pub fn empty() -> Self {
        Self {
            color: Color::WHITE,
            min_size: Vec2::ZERO,
        }
    }

    pub fn sized(color: Color, size: Vec2) -> Self {
        Self {
            color,
            min_size: size,
        }
    }

    pub fn container(color: Color) -> Self {
        Self {
            color,
            min_size: Vec2::ZERO,
        }
    }

    pub fn show(self) -> Response<ColoredBoxWidget> {
        widget::<ColoredBoxWidget>(self)
    }

    pub fn show_children<F: FnOnce()>(self, children: F) -> Response<ColoredBoxWidget> {
        widget_children::<ColoredBoxWidget, F>(children, self)
    }
}

#[derive(Debug)]
pub struct ColoredBoxWidget {
    props: ColoredBox,
}

pub type ColoredBoxResponse = ();

impl Widget for ColoredBoxWidget {
    type Props = ColoredBox;
    type Response = ColoredBoxResponse;

    fn new() -> Self {
        Self {
            props: ColoredBox::empty(),
        }
    }

    fn update(&mut self, props: Self::Props) -> Self::Response {
        self.props = props;
    }

    fn layout(&self, ctx: LayoutContext<'_>, input: Constraints) -> Vec2 {
        let node = ctx.dom.get_current();
        let mut size = self.props.min_size;

        for &child in &node.children {
            let child_size = ctx.layout.calculate(ctx.dom, child, input);
            size = size.max(child_size);
        }

        input.constrain_min(size)
    }

    fn paint(&self, mut ctx: PaintContext<'_>) {
        let node = ctx.dom.get_current();
        let layout_node = ctx.layout.get(ctx.dom.current()).unwrap();

        let mut rect = PaintRect::new(layout_node.rect);
        rect.color = self.props.color;
        ctx.paint.add_rect(rect);

        for &child in &node.children {
            ctx.paint(child);
        }
    }
}
