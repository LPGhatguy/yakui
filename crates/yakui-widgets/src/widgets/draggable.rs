use yakui_core::event::{EventInterest, EventResponse, WidgetEvent};
use yakui_core::geometry::Vec2;
use yakui_core::input::MouseButton;
use yakui_core::widget::{EventContext, Widget};
use yakui_core::Response;

use crate::util::widget_children;

#[derive(Debug)]
#[non_exhaustive]
pub struct Draggable {}

impl Draggable {
    pub fn new() -> Self {
        Draggable {}
    }

    pub fn show<F: FnOnce()>(self, children: F) -> Response<DraggableWidget> {
        widget_children::<DraggableWidget, F>(children, self)
    }
}

#[derive(Debug)]
#[non_exhaustive]
pub struct DraggableWidget {
    held: bool,
    down_position: Vec2,
    mouse_position: Vec2,
}

#[non_exhaustive]
pub struct DraggableResponse {
    pub dragging: Option<Dragging>,
}

#[derive(Debug)]
pub struct Dragging {
    pub start: Vec2,
    pub current: Vec2,
}

impl Widget for DraggableWidget {
    type Props = Draggable;
    type Response = DraggableResponse;

    fn new() -> Self {
        Self {
            held: false,
            down_position: Vec2::ZERO,
            mouse_position: Vec2::ZERO,
        }
    }

    fn update(&mut self, _props: Self::Props) -> Self::Response {
        let dragging = if self.held {
            Some(Dragging {
                start: self.down_position,
                current: self.mouse_position,
            })
        } else {
            None
        };

        DraggableResponse { dragging }
    }

    fn event_interest(&self) -> EventInterest {
        EventInterest::MOUSE_ALL
    }

    fn event(&mut self, _ctx: EventContext<'_>, event: &WidgetEvent) -> EventResponse {
        match *event {
            WidgetEvent::MouseButtonChanged {
                button: MouseButton::One,
                down,
                inside,
                position,
                ..
            } => {
                if down && inside {
                    self.held = true;
                    self.down_position = position;
                    self.mouse_position = position;
                    EventResponse::Sink
                } else if !down && self.held {
                    self.held = false;
                    EventResponse::Sink
                } else {
                    EventResponse::Bubble
                }
            }
            WidgetEvent::MouseMoved(Some(pos)) => {
                if self.held {
                    self.mouse_position = pos;
                }

                EventResponse::Bubble
            }
            _ => EventResponse::Bubble,
        }
    }
}
