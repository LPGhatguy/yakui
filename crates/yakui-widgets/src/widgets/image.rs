use yakui_core::geometry::{Color, Constraints, Rect, Vec2};
use yakui_core::paint::PaintRect;
use yakui_core::widget::{LayoutContext, PaintContext, Widget};
use yakui_core::{ManagedTextureId, Response};

use crate::util::widget;

/**
Displays an image.

Responds with [ImageResponse].
*/
#[derive(Debug, Clone)]
#[non_exhaustive]
pub struct Image {
    pub image: Option<ManagedTextureId>,
    pub size: Vec2,
}

impl Image {
    pub fn new(image: ManagedTextureId, size: Vec2) -> Self {
        Self {
            image: Some(image),
            size,
        }
    }

    pub fn show(self) -> Response<ImageWidget> {
        widget::<ImageWidget>(self)
    }
}

#[derive(Debug)]
pub struct ImageWidget {
    props: Image,
}

pub type ImageResponse = ();

impl Widget for ImageWidget {
    type Props = Image;
    type Response = ImageResponse;

    fn new() -> Self {
        Self {
            props: Image {
                image: None,
                size: Vec2::ZERO,
            },
        }
    }

    fn update(&mut self, props: Self::Props) -> Self::Response {
        self.props = props;
    }

    fn layout(&self, _ctx: LayoutContext<'_>, input: Constraints) -> Vec2 {
        input.constrain_min(self.props.size)
    }

    fn paint(&self, ctx: PaintContext<'_>) {
        let layout_node = ctx.layout.get(ctx.dom.current()).unwrap();

        if let Some(image) = self.props.image {
            let mut rect = PaintRect::new(layout_node.rect);
            rect.color = Color::WHITE;
            rect.texture = Some((image, Rect::ONE));
            ctx.paint.add_rect(rect);
        }
    }
}
