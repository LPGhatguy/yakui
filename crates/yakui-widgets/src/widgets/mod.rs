mod align;
mod button;
mod checkbox;
mod colored_box;
mod constrained_box;
mod draggable;
mod flexible;
mod image;
mod list;
mod nineslice;
mod offset;
mod pad;
mod panel;
mod render_text;
mod render_textbox;
mod scrollable;
mod state;
mod text;
mod textbox;
mod unconstrained_box;
mod window;

pub use self::align::*;
pub use self::button::*;
pub use self::checkbox::*;
pub use self::colored_box::*;
pub use self::constrained_box::*;
pub use self::draggable::*;
pub use self::flexible::*;
pub use self::image::*;
pub use self::list::*;
pub use self::nineslice::*;
pub use self::offset::*;
pub use self::pad::*;
pub use self::panel::*;
pub use self::render_text::*;
pub use self::render_textbox::*;
pub use self::scrollable::*;
pub use self::state::*;
pub use self::text::*;
pub use self::textbox::*;
pub use self::unconstrained_box::*;
pub use self::window::*;
