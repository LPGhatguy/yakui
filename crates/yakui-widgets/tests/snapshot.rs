use yakui::{Constraints, CrossAxisAlignment, MainAxisAlignment, MainAxisSize, Vec2};
use yakui_core::geometry::Color;
use yakui_core::Alignment;
use yakui_test::{run, Test};
use yakui_widgets::widgets::{Button, List, Pad, UnconstrainedBox};
use yakui_widgets::{
    align, button, center, checkbox, colored_box, colored_box_container, column, constrained,
    expanded, pad, row, text,
};

#[test]
fn colored_box_basic() {
    run!({
        colored_box(Color::RED, [50.0, 50.0]);
    });
}

#[test]
fn column_basic() {
    run!({
        column(|| {
            rect_50x50();
            rect_50x50();
            rect_50x50();
        });
    });
}

#[test]
fn row_basic() {
    run!({
        row(|| {
            rect_50x50();
            rect_50x50();
            rect_50x50();
        });
    });
}

#[test]
fn column_main_align_end() {
    run!({
        let mut container = List::column();
        container.main_axis_alignment = MainAxisAlignment::End;
        container.show(|| {
            rect_50x50();
            rect_50x50();
            rect_50x50();
        });
    });
}

#[test]
fn column_main_align_center() {
    run!({
        let mut container = List::column();
        container.main_axis_alignment = MainAxisAlignment::Center;
        container.show(|| {
            rect_50x50();
            rect_50x50();
            rect_50x50();
        });
    });
}

#[test]
fn column_item_spacing() {
    run!({
        let mut container = List::column();
        container.item_spacing = 10.0;
        container.show(|| {
            rect_50x50();
            rect_50x50();
            rect_50x50();
        });
    });
}

#[test]
fn row_item_spacing() {
    run!({
        let mut container = List::row();
        container.item_spacing = 10.0;
        container.show(|| {
            rect_50x50();
            rect_50x50();
            rect_50x50();
        });
    });
}

#[test]
fn row_grow_first() {
    run!({
        row(|| {
            expanded(|| {
                colored_box(Color::RED, [50.0, 50.0]);
            });
            colored_box(Color::GREEN, [50.0, 50.0]);
            colored_box(Color::BLUE, [50.0, 50.0]);
        });
    });
}

#[test]
fn row_grow_middle() {
    run!({
        row(|| {
            colored_box(Color::RED, [50.0, 50.0]);
            expanded(|| {
                colored_box(Color::GREEN, [50.0, 50.0]);
            });
            colored_box(Color::BLUE, [50.0, 50.0]);
        });
    });
}

#[test]
fn row_grow_last() {
    run!({
        row(|| {
            colored_box(Color::RED, [50.0, 50.0]);
            colored_box(Color::GREEN, [50.0, 50.0]);
            expanded(|| {
                colored_box(Color::BLUE, [50.0, 50.0]);
            });
        });
    });
}

#[test]
fn row_grow_middle_spacing() {
    run!({
        let mut container = List::row();
        container.item_spacing = 10.0;
        container.show(|| {
            colored_box(Color::RED, [50.0, 50.0]);
            expanded(|| {
                colored_box(Color::GREEN, [50.0, 50.0]);
            });
            colored_box(Color::BLUE, [50.0, 50.0]);
        });
    });
}

#[test]
fn column_grow_middle() {
    run!({
        column(|| {
            colored_box(Color::RED, [50.0, 50.0]);
            expanded(|| {
                colored_box(Color::GREEN, [50.0, 50.0]);
            });
            colored_box(Color::BLUE, [50.0, 50.0]);
        });
    });
}

#[test]
fn column_min() {
    run!({
        let mut column = List::column();
        column.main_axis_size = MainAxisSize::Min;
        column.show(|| {
            rect_50x50();
            rect(100, 50);
            rect(50, 100);
        });
    });
}

#[test]
fn column_cross_stretch() {
    run!({
        let mut column = List::column();
        column.cross_axis_alignment = CrossAxisAlignment::Stretch;
        column.show(|| {
            colored_box(Color::RED, [50.0, 50.0]);
            colored_box(Color::GREEN, [50.0, 50.0]);
            colored_box(Color::BLUE, [50.0, 50.0]);
        });
    });
}

#[test]
fn row_cross_start() {
    run!({
        let mut row = List::row();
        row.cross_axis_alignment = CrossAxisAlignment::Start;
        row.show(|| {
            colored_box(Color::RED, [50.0, 50.0]);
            colored_box(Color::GREEN, [50.0, 50.0]);
            colored_box(Color::BLUE, [50.0, 50.0]);
        });
    });
}

#[test]
fn row_cross_center() {
    run!({
        let mut row = List::row();
        row.cross_axis_alignment = CrossAxisAlignment::Center;
        row.show(|| {
            colored_box(Color::RED, [50.0, 50.0]);
            colored_box(Color::GREEN, [50.0, 50.0]);
            colored_box(Color::BLUE, [50.0, 50.0]);
        });
    });
}

#[test]
fn row_cross_end() {
    run!({
        let mut row = List::row();
        row.cross_axis_alignment = CrossAxisAlignment::End;
        row.show(|| {
            colored_box(Color::RED, [50.0, 50.0]);
            colored_box(Color::GREEN, [50.0, 50.0]);
            colored_box(Color::BLUE, [50.0, 50.0]);
        });
    });
}

#[test]
fn row_cross_stretch() {
    run!({
        let mut row = List::row();
        row.cross_axis_alignment = CrossAxisAlignment::Stretch;
        row.show(|| {
            colored_box(Color::RED, [50.0, 50.0]);
            colored_box(Color::GREEN, [50.0, 50.0]);
            colored_box(Color::BLUE, [50.0, 50.0]);
        });
    });
}

/// When given infinite constraints, widgets like List need to pick the minimum
/// size that fits their content, not infinity.
#[test]
fn row_unconstrained() {
    run!({
        let container = UnconstrainedBox::new();
        container.show(|| {
            row(|| {
                colored_box(Color::RED, [50.0, 50.0]);
                colored_box(Color::GREEN, [50.0, 50.0]);
                colored_box(Color::BLUE, [50.0, 50.0]);
            });
        });
    });
}

#[test]
fn align_center() {
    run!({
        center(|| {
            rect_50x50();
        });
    });
}

#[test]
fn align_text_center() {
    run!({
        center(|| {
            text(60.0, "X X X X X");
        });
    });
}

#[test]
fn align_bottom_right() {
    run!({
        align(Alignment::BOTTOM_RIGHT, || {
            rect_50x50();
        });
    });
}

#[test]
fn pad_basic() {
    let padding = Pad::all(20.0);

    run!({
        align(Alignment::TOP_LEFT, || {
            colored_box_container(Color::RED, || {
                pad(padding, || {
                    colored_box(Color::GREEN, Vec2::splat(50.0));
                });
            });
        });
    });
}

#[test]
fn pad_empty() {
    let padding = Pad::all(20.0);

    run!({
        align(Alignment::TOP_LEFT, || {
            colored_box_container(Color::RED, || {
                pad(padding, || {});
            });
        });
    });
}

#[test]
fn button_text_alignment() {
    run!({
        let constraints = Constraints::tight(Vec2::splat(800.0));
        align(Alignment::TOP_LEFT, || {
            constrained(constraints, || {
                let mut button = Button::styled("X X X X X".into());
                button.text_style.font_size = 60.0;
                button.show();
            });
        });
    });
}

#[test]
fn constrain_checkbox() {
    run!({
        let constraints = Constraints::tight(Vec2::splat(100.0));

        center(|| {
            constrained(constraints, || {
                checkbox(true);
            });
        });
    });
}

#[test]
fn column_intrinsic() {
    run!({
        column(|| {
            button("X X X");
        });
    });
}

fn rect<V: Dim>(w: V, h: V) {
    colored_box(Color::WHITE, [w.to_f32(), h.to_f32()]);
}

fn rect_50x50() {
    rect(50.0, 50.0);
}

trait Dim {
    fn to_f32(self) -> f32;
}

impl Dim for f32 {
    fn to_f32(self) -> f32 {
        self
    }
}

impl Dim for i32 {
    fn to_f32(self) -> f32 {
        self as f32
    }
}
