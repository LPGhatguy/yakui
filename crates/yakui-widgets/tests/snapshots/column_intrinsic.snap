---
source: crates/yakui-widgets/tests/snapshot.rs
assertion_line: 337
expression: view
---
- ListWidget pos(0, 0) size(1000, 1000)
  - ButtonWidget pos(0, 0) size(74, 36)
    - RoundRectWidget pos(0, 0) size(74, 36)
      - PadWidget pos(0, 0) size(74, 36)
        - AlignWidget pos(20, 10) size(34, 16)
          - RenderTextWidget pos(20, 10) size(34, 16)
